cpp/src/arrow/flight/client.cc

```diff
@@ -993,6 +1010,36 @@ class FlightClient::FlightClientImpl {
     return Status::OK();
   }
 
+  Status AuthenticateBasicToken(const std::string& username, const std::string& password,
+                                std::pair<std::string, std::string>* bearer_token) {
+    // Add bearer token factory to middleware so it can intercept the bearer token.
+    if (interceptor_pointer != NULLPTR) {
```

Comment:
Honestly, instead of using interceptors, why not use ClientRpc.context (a [grpc::ClientContext](https://grpc.github.io/grpc/cpp/classgrpc_1_1_client_context.html)) to attach the headers (AddMetadata) and retrieve them (GetServerInitialMetadata)? This isn't a general auth mechanism anyways, so I'd rather we keep the implementation simple & tightly scoped in this case.